### PR TITLE
Faster SubMatch implementation

### PR DESCRIPTION
**Pull request by DavidObando**
SubMatch has been sped up by implementing a modified Boyer–Moore–Horspool algorithm with an average-case complexity of O(N) on random text. Worst case, it behaves similarly to the previous implementation O(MN), where M is the length of the boundary and N is the length of the buffer to operate on.

Method SubMatch looks for two things:

  1. Whether the byte array segment fully contains the boundary, or
  2. Whether the byte array ends with the start of the boundary.

Case 1 is now a lot faster than the previous implementation. Case 2 remains using the same code as before. The method will do Case 1 until the matchOffset is equal to N-M. It then switches to Case 2, unless a match is found.

The code can be further sped up with a full Boyer–Moore implementation, or something more sophisticated. This however can be evaluated in the case that this implementation is insufficiently performant for our main scenarios.

This commit resolves issue #575.

Using test [MultipartPOST](https://github.com/aspnet/Performance/tree/dev/testapp/MultipartPOST) the work percentage (inclusive %) for method SubMatch under the profiler was:
  * Before this change: 53.4%
  * After this change: 7.6%

@SajayAntony @Tratcher @pranavkm @davidfowl

### CODE REVIEW

File: src/Microsoft.AspNetCore.WebUtilities/MultipartReaderStream.cs [hunk 1]
```diff
@@ -40,15 +38,8 @@ public MultipartReaderStream(BufferedReadStream stream, string boundary, bool ex
...
-            {
-                _boundaryBytes = Encoding.UTF8.GetBytes("\r\n--" + boundary);
-            }
-            else
-            {
-                _boundaryBytes = Encoding.UTF8.GetBytes("--" + boundary);
-            }
-            _finalBoundaryLength = _boundaryBytes.Length + 2; // Include the final '--' terminator.
+            _boundary = boundary;
+            _boundary.ExpectLeadingCrlf = expectLeadingCrlf;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/MultipartReaderStream.cs]: It's weird to take a constructor parameter that only modifies another parameter, especially one that already takes the same parameter.

Comment (davidfowl) [inline src/Microsoft.AspNetCore.WebUtilities/MultipartReaderStream.cs]: Yea it should be set from the caller on the multipart boundary object

Comment (SajayAntony) [inline src/Microsoft.AspNetCore.WebUtilities/MultipartReaderStream.cs]: I can't understand why is there an optional parameter. But with your change it doesn't make sense to pass it in anymore since it can be set on the `MultipartBoundary` object. 
@Tratcher @davidfowl couldn't we just remove this now and set it at the call site as mentioned.

Comment (DavidObando) [inline src/Microsoft.AspNetCore.WebUtilities/MultipartReaderStream.cs]: Yup, I failed to see that. Will send an update to fix this.

**Comment by SajayAntony** [comment]
Awesome drop in CPU.

**Comment by DavidObando** [comment]
Just updated this PR to address feedback. Let me know if you see anything else. Thanks!

**Comment by DavidObando** [comment]
Ping @SajayAntony @Tratcher @pranavkm @davidfowl

**Comment by davidfowl** [comment]
Do we have any tests for this code? Unit tests would be best

**Comment by DavidObando** [comment]
@davidfowl we do have tests for MultipartReader, and they seem adequate. Do you have any scenario in mind, I'd gladly add it.

**Comment by davidfowl** [comment]
@DavidObando test cases for each of the cases called out in the algorithm.

**Comment by DavidObando** [comment]
@davidfowl Just sent an update with a test. The only missing case was testing that a partial boundary at the end of the buffer still signals the end of a section, so I added a test for that (effectively exercising case 2). Case 1 is fully covered by this and other tests.

**Comment by davidfowl** [comment]
Nice!

File: test/Microsoft.AspNetCore.WebUtilities.Tests/MultipartReaderTests.cs [hunk 2]
```diff
@@ -218,6 +239,47 @@ private static MemoryStream MakeStream(string text)
...
+            var stream = MakeStream(TwoPartBodyIncompleteBuffer);
+            var reader = new MultipartReader(Boundary, stream);
+            var buffer = new byte[128];
+
+            //first section can be read successfully
+            var section = await reader.ReadNextSectionAsync();
+            Assert.NotNull(section);
+            Assert.Equal(1, section.Headers.Count);
+            Assert.Equal("form-data; name=\"text\"", section.Headers["Content-Disposition"][0]);
+            section.Body.Read(buffer, 0, buffer.Length);
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/MultipartReaderTests.cs]: why not use the value returned from read in `GetStringASCIIZ`?

**Comment by davidfowl** [comment]
:shipit:

File: test/Microsoft.AspNetCore.WebUtilities.Tests/MultipartReaderTests.cs (hunk 2 above)
Comment (DavidObando) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/MultipartReaderTests.cs]: @davidfowl Yes, thanks for the observation. Fixed now.

**Comment by DavidObando** [comment]
Merged 15649b7.
